### PR TITLE
DBA-361: Vitess MVP

### PR DESCRIPTION
**Pull request by artemvovk**
the most basic setup to at least somewhat get Vitess events from VTGate gRPC and shove them into a producer

# For reference:

1. VTGate protos: https://github.com/vitessio/vitess/blob/main/proto/binlogdata.proto
2. Query/Row protos: https://github.com/vitessio/vitess/blob/main/proto/query.proto
3. Debezium VTGate connection code: https://github.com/debezium/debezium-connector-vitess/tree/main/src/main/java/io/debezium/connector/vitess/connection
4. Vitess's Java code: https://github.com/vitessio/vitess/tree/main/java

### Does NOT

- track schemas
- handle resharding

## Logs:

Screen Shot 2021-08-03 at 10 50 22

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -123,7 +123,17 @@
       </dependencies>
     </profile>
   </profiles>
-
+  <dependencyManagement>
```
Comment (artemvovk) [inline pom.xml:126]: fixes conflicts between BQ transitive dependencies and gRPC's transitive dependencies

File: src/main/java/com/zendesk/maxwell/schema/columndef/DateTimeColumnDef.java [hunk 2]
```diff
@@ -27,6 +27,8 @@ protected String formatValue(Object value, MaxwellOutputConfig config) throws Co
 				else
 					return appendFractionalSeconds("0000-00-00 00:00:00", 0, getColumnLength());
 			}
+		} if ( value instanceof String ) {
```
Comment (artemvovk) [inline src/main/java/com/zendesk/maxwell/schema/columndef/DateTimeColumnDef.java:30]: could add some validation here... if it ever bugs out

File: src/main/java/com/zendesk/maxwell/schema/columndef/IntColumnDef.java [hunk 3]
```diff
@@ -58,11 +58,20 @@ private Long toLong(Object value) throws ColumnDefCastException {
 
 			long res = castUnsigned(i, 1L << this.bits);
 			return Long.valueOf(res);
+		} else if (value instanceof String) {
```
Comment (artemvovk) [inline src/main/java/com/zendesk/maxwell/schema/columndef/IntColumnDef.java:61]: life is hard

File: src/main/java/com/zendesk/maxwell/replication/VitessConnectorReplicator.java [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+				String columnType = field.getType().toString().toLowerCase(Locale.ROOT);
+				String[] enumValues = new String[0];
+				if (columnType.contains("int")) {
+					columnType = "int";
+				} else if (columnType.contains("float")) {
+					columnType = "float";
+				} else if (columnType.contains("enum")) {
+					String ct = field.getColumnType();
+					String[] split = ct.split("\\'");
+					enumValues = Stream.of(split)
```
Comment (artemvovk) [inline src/main/java/com/zendesk/maxwell/replication/VitessConnectorReplicator.java:291]: yes this is terrible and.... life is hard

File: src/main/java/com/zendesk/maxwell/MaxwellConfig.java [hunk 5]
```diff
@@ -549,12 +552,14 @@ private void parse(String [] argv) {
 	}
 
 	private void setup(OptionSet options, Properties properties) {
-		this.log_level = fetchStringOption("log_level", options, properties, null);
+		this.log_level = fetchStringOption("log_level", options, properties, "INFO");
```
Comment (rpedela-recurly) [inline src/main/java/com/zendesk/maxwell/MaxwellConfig.java]: whats the difference between null and info?

File: src/main/java/com/zendesk/maxwell/replication/VitessConnectorReplicator.java [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+							.filter(Predicate.not(value -> value.startsWith("enum(")))
+							.filter(Predicate.not(value -> value.equals(",")))
+							.filter(Predicate.not(value -> value.equals(")")))
+							.collect(Collectors.toSet()).toArray(new String[0]);
+				}
+				ColumnDef columnDef = ColumnDef.build(
```
Comment (rpedela-recurly) [inline src/main/java/com/zendesk/maxwell/replication/VitessConnectorReplicator.java:297]: What is different/special about int, float, enum that `ColumnDef.build()` can't handle them directly?

File: src/main/java/com/zendesk/maxwell/schema/columndef/DateTimeColumnDef.java (hunk 2 above)
Comment (rpedela-recurly) [inline src/main/java/com/zendesk/maxwell/schema/columndef/DateTimeColumnDef.java:30]: What is the data format case this is handling? Do you have an example?

**Review by rpedela-recurly — COMMENTED** [review]
Great start!

File: src/main/java/com/zendesk/maxwell/MaxwellConfig.java (hunk 5 above)
Comment (artemvovk) [reply src/main/java/com/zendesk/maxwell/MaxwellConfig.java]: derp, from testing, will undo

File: src/main/java/com/zendesk/maxwell/replication/VitessConnectorReplicator.java (hunk 6 above)
Comment (artemvovk) [reply src/main/java/com/zendesk/maxwell/replication/VitessConnectorReplicator.java:297]: https://github.com/vitessio/vitess/blob/main/proto/query.proto#L110

the contract defines them as Golang types, not actual mysql types - so we have to "derive" these. There's a jira ticket to make this more stable and cleaner looking

File: src/main/java/com/zendesk/maxwell/schema/columndef/DateTimeColumnDef.java (hunk 2 above)
Comment (artemvovk) [reply src/main/java/com/zendesk/maxwell/schema/columndef/DateTimeColumnDef.java:30]: So basically we get all `values` in one string and then we chop it up into substrings. So this is how we have to "derive" the actual Java type out of that chopped up string.
Screen Shot 2021-08-04 at 14 51 30

**Pull request merged**